source/agent/addons/quic/QuicTransportStream.cc

```diff
@@ -0,0 +1,225 @@
+}
+
+NAN_MODULE_INIT(QuicTransportStream::init)
+{
+    Local<FunctionTemplate> tpl = Nan::New<FunctionTemplate>(newInstance);
+    tpl->SetClassName(Nan::New("QuicTransportStream").ToLocalChecked());
+    Local<ObjectTemplate> instanceTpl = tpl->InstanceTemplate();
+    instanceTpl->SetInternalFieldCount(1);
+
+    Nan::SetPrototypeMethod(tpl, "write", write);
+    Nan::SetPrototypeMethod(tpl, "addDestination", addDestination);
+    //Nan::SetPrototypeMethod(tpl, "write", write);
+
+    s_constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
+    Nan::Set(target, Nan::New("QuicTransportStream").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
+}
+
+NAN_METHOD(QuicTransportStream::newInstance)
+{
+    if (!info.IsConstructCall()) {
+        ELOG_DEBUG("Not construct call.");
+        return;
+    }
+    QuicTransportStream* obj = new QuicTransportStream();
+    obj->Wrap(info.This());
+    uv_async_init(uv_default_loop(), &obj->m_asyncOnContentSessionId, &QuicTransportStream::onContentSessionId);
+    uv_async_init(uv_default_loop(), &obj->m_asyncOnData, &QuicTransportStream::onData);
+    info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(QuicTransportStream::write){
+    if(info.Length()<2){
+        Nan::ThrowTypeError("Data is not provided.");
+        return;
+    }
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info.Holder());
+    uint8_t* buffer=(uint8_t*)node::Buffer::Data(info[0]->ToObject());
+    unsigned int length = info[1]->Uint32Value();
+    obj->m_stream->Write(buffer, length);
+    info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(QuicTransportStream::addDestination)
+{
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info.Holder());
+    if (info.Length() != 1) {
+        Nan::ThrowTypeError("Invalid argument length for addDestination.");
+        return;
+    }
+    // TODO: Check if info[0] is an Nan wrapped object.
+    auto framePtr = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info[0]->ToObject());
+    // void* ptr = info[0]->ToObject()->GetAlignedPointerFromInternalField(0);
+    // auto framePtr=static_cast<owt_base::FrameDestination*>(ptr);
+    obj->addDataDestination(framePtr);
+    obj->m_isPiped = true;
+}
+
+NAN_METHOD(QuicTransportStream::removeDestination)
+{
+}
+
+v8::Local<v8::Object> QuicTransportStream::newInstance(owt::quic::QuicTransportStreamInterface* stream)
+{
+    Local<Object> streamObject = Nan::NewInstance(Nan::New(QuicTransportStream::s_constructor)).ToLocalChecked();
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(streamObject);
+    obj->m_stream = stream;
+    obj->MaybeReadContentSessionId();
+    return streamObject;
+}
+
+void QuicTransportStream::MaybeReadContentSessionId()
+{
+    if (!m_receivedContentSessionId) {
+        // Match to a content session.
+        if (m_stream->ReadableBytes() > 0 && m_stream->ReadableBytes() < uuidSizeInByte) {
```

Comment:
I think it's dangerous to assume all 16 bytes coming together and otherwise close the stream.  I would suggest all the logic of checking sessionID goes to onData as that's the supposed single entry point of receiving data.

Reply:
Yes, this is a problem. JavaScript code can handle this case, but the code here assumes all 16 bytes coming together. Perhaps filing a bug first.